### PR TITLE
Refactorings

### PR DESCRIPTION
**Pull request by vinjana**
More robust and concise dealing with control and tumor BAMs. More tests.

### CODE REVIEW

File: src/de/dkfz/b080/co/common/BasicCOProjectsRuntimeService.groovy [hunk 1]
```diff
@@ -251,11 +299,11 @@ class BasicCOProjectsRuntimeService extends RuntimeService {
         return new File(getInpDirectory(COConstants.CVALUE_SEQUENCE_DIRECTORY, process, sample, library).getAbsolutePath().replace('${run}', run));
     }
 
-    BasicBamFile getMergedBamFileForDataSetAndSample(ExecutionContext context, Sample sample) {
-        return getMergedBamFileForDataSetAndSample(context, null, sample);
+    BasicBamFile getMergedBamFileFromFilesystem(ExecutionContext context, Sample sample) {
```
Comment (dankwart-de) [inline src/de/dkfz/b080/co/common/BasicCOProjectsRuntimeService.groovy]: Remove this method and set dataSet to null by default in the next method?

File: src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy [hunk 2]
```diff
@@ -64,59 +64,45 @@ abstract class WorkflowUsingMergedBams extends Workflow {
...
-                    for (Sample sample : samples)
-                        assignBam(runtimeService.getMergedBamFileForDataSetAndSample(context, sample))
-
+            // Check the cache for the dataset. If not, try to load the files.
+            if (!foundBamFilesForDatasets.containsKey(dataSet)) {
+
+                runtimeService.getAllBamFiles(context).collect { BasicBamFile bam ->
+                    Sample sample = bam.sample
+                    if (sample.getType() == Sample.SampleType.CONTROL)
+                        bamsControlMerged.add(bam)
```
Comment (dankwart-de) [inline src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy]: .add => <<

File: src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy [hunk 3]
```diff
@@ -64,59 +64,45 @@ abstract class WorkflowUsingMergedBams extends Workflow {
...
+                    else if (sample.getType() == Sample.SampleType.TUMOR)
+                        bamsTumorMerged.add(bam)
```
Comment (dankwart-de) [inline src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy]: dito

File: src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy [hunk 4]
```diff
@@ -64,59 +64,45 @@ abstract class WorkflowUsingMergedBams extends Workflow {
...
+
+                if (bamsControlMerged.size() > 1) {
+                    context.addErrorEntry(ExecutionContextError.EXECUTION_SETUP_INVALID.expand("More than one control sample in bamfile_list:\n" +
+                            bamsControlMerged.collect { "${it.sample.name}: ${it.getAbsolutePath()}" }.join("\n")))
                 }
-                if (!isNoControlWorkflow(context) && bamControlMerged) allFound.add(bamControlMerged)
-                allFound.addAll(bamsTumorMerged)
-                foundBamFilesForDatasets.put(dataSet, allFound as BasicBamFile[])
+                if (!isNoControlWorkflow(context) && bamsControlMerged.size() == 1)
+                    foundBamFilesForDatasets.put(dataSet, (bamsControlMerged + bamsTumorMerged) as BasicBamFile[])
```
Comment (dankwart-de) [inline src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy]: .put(dataSet => [dataSet] =

File: src/de/dkfz/b080/co/common/BasicCOProjectsRuntimeService.groovy (hunk 1 above)
Comment (vinjana) [reply src/de/dkfz/b080/co/common/BasicCOProjectsRuntimeService.groovy]: O.k. Note, however, I still dislike null -- also as parameter. The only way to figure out, whether a method accepts null for a parameter value is trial and error or reading the code. Both significantly hamper development speed. Better is to have the interface express this. Either by not having the parameter, or by using Optional.

File: src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy (hunk 2 above)
Comment (vinjana) [reply src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy]: done

File: src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy (hunk 3 above)
Comment (vinjana) [reply src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy]: done

File: src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy (hunk 4 above)
Comment (vinjana) [reply src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy]: done

**Pull request merged**